internal/target/service_list_refresh.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package target
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+// ListRefresh lists targets according to the page size
+// and refresh token, filtering out entries that do not
+// pass the filter item fn. It returns a new refresh token
+// based on the old one, the grants hash, and the returned
+// targets.
+func ListRefresh(
+	ctx context.Context,
+	tok *refreshtoken.Token,
+	repo *Repository,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn func(Target) (bool, error),
+) (*ListResponse, error) {
+	const op = "target.ListRefresh"
+
+	deletedIds, transactionTimestamp, err := repo.listDeletedIds(ctx, tok.UpdatedTime)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	limit := pageSize + 1
+	opts := []Option{
+		WithLimit(limit),
+		WithStartPageAfterItem(tok.LastItemId, tok.LastItemUpdatedTime),
+	}
+
+	targets := make([]Target, 0, pageSize+1)
+dbLoop:
+	for {
+		// Request another page from the DB until we fill the final items
+		page, err := repo.ListTargets(ctx, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		for _, item := range page {
+			ok, err := filterItemFn(item)
+			if err != nil {
+				return nil, errors.Wrap(ctx, err, op)
+			}
+			if ok {
+				targets = append(targets, item)
+				// If we filled the items after filtering,
+				// we're done.
+				if len(targets) == cap(targets) {
+					break dbLoop
+				}
+			}
+		}
+		// If the current page was shorter than the limit, stop iterating
+		if len(page) < limit {
+			break dbLoop
+		}
+
+		opts = []Option{
+			WithLimit(limit),
+			WithStartPageAfterItem(page[len(page)-1].GetPublicId(), page[len(page)-1].GetUpdateTime().AsTime()),
+		}
+	}
+	// If we couldn't fill the items, it was a complete listing.
+	completeListing := len(targets) < cap(targets)
+	if !completeListing {
+		// Items is of size pageSize+1, so
+		// truncate if it was filled.
+		targets = targets[:pageSize]
+	}
+
+	totalItems, err := repo.estimatedCount(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	resp := &ListResponse{
+		Items:               targets,
+		DeletedIds:          deletedIds,
+		EstimatedTotalItems: totalItems,
+		CompleteListing:     completeListing,
+	}
+
+	// Use the timestamp of the deleted IDs transaction with a
+	// buffer to account for overlapping transactions. It is okay
+	// to return a deleted ID more than once. The buffer corresponds
+	// to Postgres' default transaction timeout.
+	updatedTime := transactionTimestamp.Add(-30 * time.Second)
+	if updatedTime.Before(tok.CreatedTime) {
+		// Ensure updated time isn't before created time due
+		// to the buffer.
+		updatedTime = tok.CreatedTime
+	}
```

Comment:
@jimlambrt @tmessi note this truncation of the updated time to at least the timestamp of the created time. It means much of our planned buffer disappears when a user performs the initial paginating-of-all-results, but hopefully it shouldn't matter since the number of deleted IDs since their tokens were created should also be small.

Reply:
I know this is already merged, but maybe as a follow up, I think this logic should be in a method on the RefreshToken, i.e. the RefreshToken should ensure that it's updated time is never before it's creation time. I also wonder if the buffer time should be something defined in the refreshtoken package. I am thinking something like:

```go
package refreshtoken

const DefaultUpdateTimeBuffer = -30 * time.Second


func (r *RefreshToken) SetUpdateTime(u time.Time, opts ...Option) {
  // opts stuff

  uu = u.Add(opts.withUpdateTimeBuffer)
  if uu.Before(r.CreatedTime) {
     r.UpdatedTime = r.CreatedTime
  } else {
     r.UpdatedTime = uu
  }
}
```

Reply:
Fair point, I'll make a separate PR for this.